aws/resource_aws_sns_application_test.go

```diff
@@ -0,0 +1,204 @@
+			}
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSSNSApplicationDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSSNSApplicationAPNSSandBoxConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "name", "aws_sns_application_test"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "platform", "APNS_SANDBOX"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "success_sample_rate", "100"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "created_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-created-topic"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "updated_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-updated-topic"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "failure_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-failure-topic"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "deleted_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-deleted-topic"),
+				),
+			},
+			resource.TestStep{
+				Config: testAccAWSSNSApplicationAPNSSandBoxConfigUpdate,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "name", "aws_sns_application_test"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "platform", "APNS_SANDBOX"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "success_sample_rate", "99"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "created_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-created-topic-update"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "updated_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-updated-topic-update"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "failure_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-failure-topic-update"),
+					resource.TestCheckResourceAttr(
+						"aws_sns_application.apns_test", "deleted_topic", "arn:aws:sns:us-east-1:638386993804:endpoint-deleted-topic-update"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAWSSNSApplicationDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).snsconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_sns_application" {
+			continue
+		}
+		_, err := conn.DeletePlatformApplication(&sns.DeletePlatformApplicationInput{
+			PlatformApplicationArn: aws.String(rs.Primary.ID),
+		})
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NoSNSApplication" {
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+var testAccAWSSNSApplicationGCMConfig = `
```

Comment:
For future updates to the resource, it helps to use a function to return a string for configuration values. This also allows you to get all the necessary environment variables for a configuration prior to writing the string. 

```
func testAccAWSSNSApplicationGCMConfig(gcm_key string) string {
  return fmt.Sprintf(`foo bar baz`)
}
```